File: framework/languages/dotnet/tutorial/Server/Actors/PlayerHandlers.cs
using Tutorial.Server.Spots;
using Tutorial.Shared;
using Zlink.Framework.Contracts.Handlers;
using Zlink.Framework.Contracts.Spots;

namespace Tutorial.Server.Actors;

// A message addressed to a player runs inside the Spot the player currently
// occupies, so handlers receive both. Players in the lobby use the Entry Spot
// interfaces below; players inside a room use IZLinkSpotActor*Handler instead.

// --8<-- [start:actor-handlers]
// --8<-- [start:actor-send-handler]
public sealed class ChangeNicknameHandler
    : IZLinkEntrySpotActorSendHandler<LobbySpot, Player, ChangeNickname>
{
    public async ValueTask HandleAsync(
        LobbySpot lobby,
        Player player,
        IZLinkMessageContext context,
        ChangeNickname message,
        CancellationToken cancellationToken)
    {
        player.Rename(message.Nickname);

        // --8<-- [start:actor-push]
        // Reaches the connection bound to this player. If none is bound, the
        // call does nothing rather than failing.
        await player.Context.BoundSession
            .Send(new NicknameChanged(player.Nickname))
            .Async(cancellationToken);
        // --8<-- [end:actor-push]
    }
}
// --8<-- [end:actor-send-handler]
// --8<-- [start:actor-request-handler]

public sealed class GetPlayerHandler
    : IZLinkEntrySpotActorRequestHandler<LobbySpot, Player, GetPlayer, PlayerInfo>
{
    public ValueTask<PlayerInfo> HandleAsync(
        LobbySpot lobby,
        Player player,
        IZLinkMessageContext context,
        GetPlayer request,
        CancellationToken cancellationToken)
        => ValueTask.FromResult(new PlayerInfo(player.Context.ActorId, player.Nickname));
}
// --8<-- [end:actor-request-handler]
// --8<-- [end:actor-handlers]

File: framework/languages/java/tutorial/kotlin/Server/src/main/kotlin/systems/zlink/tutorial/server/actors/PlayerHandlers.kt
package systems.zlink.tutorial.server.actors

import kotlinx.coroutines.future.await
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorSendHandler
import systems.zlink.tutorial.server.spots.LobbySpot
import systems.zlink.tutorial.shared.ChangeNickname
import systems.zlink.tutorial.shared.GetPlayer
import systems.zlink.tutorial.shared.NicknameChanged
import systems.zlink.tutorial.shared.PlayerInfo

// A message addressed to a player runs inside the Spot the player currently
// occupies, so a handler receives both the Spot and the player.

// --8<-- [start:actor-handlers]
// --8<-- [start:actor-send-handler]
class ChangeNicknameHandler :
    ZLinkSuspendingEntrySpotActorSendHandler<LobbySpot, Player, ChangeNickname> {
    override suspend fun handle(
        entrySpot: LobbySpot,
        actor: Player,
        context: ZLinkMessageContext,
        message: ChangeNickname,
    ) {
        actor.rename(message.nickname)

        // --8<-- [start:actor-push]
        // Reaches the connection bound to this player. The same handler also runs
        // for a player nobody is connected to -- the HTTP path of the Actor step --
        // and on this binding that push fails instead of doing nothing, so the
        // failure is dropped rather than failing the rename.
        runCatching {
            actor.context().boundSession()
                .send(NicknameChanged(actor.nickname))
                .submit()
                .await()
        }
        // --8<-- [end:actor-push]
    }
}

// --8<-- [end:actor-send-handler]
// The return value is the reply. This handler only reads.
// --8<-- [start:actor-request-handler]
class GetPlayerHandler :
    ZLinkSuspendingEntrySpotActorRequestHandler<LobbySpot, Player, GetPlayer, PlayerInfo> {
    override suspend fun handle(
        entrySpot: LobbySpot,
        actor: Player,
        context: ZLinkMessageContext,
        request: GetPlayer,
    ): PlayerInfo = PlayerInfo(actor.context().actorId(), actor.nickname)
}
// --8<-- [end:actor-request-handler]
// --8<-- [end:actor-handlers]
